src/virtual_stain_flow/trainers/AbstractTrainer.py

```diff
@@ -165,37 +165,69 @@ def evaluate_step(self, inputs: torch.tensor, targets: torch.tensor)->Dict[str,
         """
         pass
     
-    @abstractmethod
-    def train_epoch(self)->dict[str, torch.Tensor]:
+    def train_epoch(self):
         """
-        Can be overridden by subclasses to implement custom training logic.
-        Make calls to the train_step method for each batch 
-        in the training DataLoader.
+        Requires implemented train_step method
+        Perform a full training epoch over the training dataset.
+        Primarily responsible for iterating over the data loader and
+            invoking train_step, and then collecting the losses.
 
-        Return a dictionary with loss name as key and 
-        torch tensor loss as value. Multiple losses can be returned.
+        Can be overridden by subclasses to implement custom training logic.
 
-        :return: A dictionary containing the loss values for the epoch.
-        :rtype: dict[str, torch.Tensor]
+        :returns: A dictionary of average loss values for the epoch.
         """
+        losses = defaultdict(list)
 
-        pass        
+        batch_idx = 0
+        for inputs, targets in self._train_loader:
 
-    @abstractmethod
-    def evaluate_epoch(self)->dict[str, torch.Tensor]:
+            self._update_epoch_progress(
+                batch_idx=batch_idx,
+                num_batches=len(self._train_loader),
+                phase="Train"
+            )
+
+            batch_loss = self.train_step(inputs, targets)
+            for key, value in batch_loss.items():
+                losses[key].append(value)
+
+            batch_idx += 1            
+
+        return {
+            key: sum(values) / len(values) for key, values in losses.items()
+        }
+    
+    def evaluate_epoch(self):
```

Comment:
Double checking: do you still want to keep the `abstractmethod` decorator here (it looked like it got scrubbed). It looked like the answer might be no given that this no longer is just a `pass` but wanted to be sure.